server/test_nlp_manual.py

```diff
@@ -31,25 +39,162 @@ def _encoded_q(query: str, embedding: str = 'us_filtered'):
+    for tile in got[0]['blocks'][0]['columns'][0]['tiles']:
+      stat_var_keys_got.update(tile['statVarKey'])
+
+    assert stat_var_keys_got == set([t.stat_var_key for t in categories.tiles])
+
+  sv_key_2_tile_dict = {}
+  for tile in got[0]['blocks'][0]['columns'][0]['tiles']:
+    for k in tile['statVarKey']:
+      sv_key_2_tile_dict[k] = tile
 
-  assert response.status_code == 200
-  assert 'config' in response.json
+  for tile in categories.tiles:
+    if tile.ranking:
+      tile_got = sv_key_2_tile_dict[tile.stat_var_key]
+      assert tile_got['type'] == "RANKING"
+      assert set([tile.ranking]) == set(tile_got['rankingTileSpec'].keys())
+
+
+def _verify_metadata(metadata: Metadata, got: Dict):
+  if metadata.place_dcid:
+    assert got.get('placeDcid') == metadata.place_dcid
+
+
+def _verify_config(config: Config, got: Dict):
+  if config.categories:
+    assert 'categories' in got
+    _verify_categories(config.categories, got['categories'])
+
+  if config.metadata:
+    assert 'metadata' in got
+    _verify_metadata(config.metadata, got['metadata'])
+
+
+@dataclass
+class QueryTestCase:
+  query: str
+  # A subset of the actual config returned.
+  # Every field(including nested fields) config_expected must be in
+  # the config returned for the test to pass.
+  config_expected: Config
+
+  def test(self, context_history: List):
+    """Tests the query by calling NL data API.
+
+    Modifies the context_history using API call result.
+    """
+    response = app.test_client().post(f'/nl/data?{_encoded_q(self.query)}',
+                                      json={'contextHistory': context_history},
+                                      content_type='application/json')
+
+    # Test NL data config matches.
+    try:
+      assert response.status_code == 200
+      assert 'config' in response.json
+
+      _verify_config(self.config_expected, response.json['config'])
+    except AssertionError:
+      print(json.dumps(response.json['context'], indent=2))
+      _print_message(f'Failed - {self.query}')
+      sys.exit('Flow failed, please see context above.')
+
+    context_history.append(response.json['context'])
+    _print_message(f"SUCCESS - '{self.query}'")
+
+
+def test_palo_alto_flow():
+  """Tests the following flow.
 
-  config = response.json['config']
-  assert 'metadata' in config and 'placeDcid' in config['metadata']
+  1. "tell me about Palo Alto"
+  2. "what about auto theft"
+  3. "what about Boston"
+  4. "what are the worst cities in California"
+  """
+  context_history = []
+  test_cases = [
+      # Q1.
+      QueryTestCase(query='tell me about Palo Alto',
+                    config_expected=Config(metadata=Metadata(
+                        place_dcid=[PALO_ALTO_DCID]))),
+      # Q2.
+      QueryTestCase(
+          query='what about auto theft',
+          config_expected=Config(
+              categories=Categories(tiles=[
+                  Tile(stat_var_key='Count_CriminalActivities_MotorVehicleTheft'
```

Comment:
I feel testing the key here would be very subject to impl details in server. Any change would require updating this test. Let's see how well it works and adjust.

Reply:
> Any change would require updating this test
Right, the downside of this approach. Curious to see how well this works as well.